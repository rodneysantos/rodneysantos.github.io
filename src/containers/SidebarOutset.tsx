import cn from "classnames";
import React, {
  createContext,
  PropsWithChildren,
  useContext,
  useState,
} from "react";

interface SidebarOutsetProps {
  value: boolean;
}

interface SidebarOutsetContextProps {
  isVisible: boolean;
  setIsVisible?: (val: boolean) => void;
}

export const SidebarOutsetContext = createContext<SidebarOutsetContextProps>({
  isVisible: false,
});

export const useSidebarOutset = () => useContext(SidebarOutsetContext);

const SidebarOutset: React.FC<SidebarOutsetProps & PropsWithChildren> = ({
  children,
  value,
}) => {
  const [isVisible, setIsVisible] = useState<boolean>(value);
  const cns = cn(
    "flex",
    "fixed",
    "font-dosis",
    "h-screen",
    "transition-transform",
    { "-translate-x-[90%]": !isVisible },

    "lg:duration-200",

    "3xl:duration-300",
  );

  return (
    <>
      <SidebarOutsetContext.Provider value={{ isVisible, setIsVisible }}>
        <div className={cns}>{children}</div>
      </SidebarOutsetContext.Provider>
    </>
  );
};

export default SidebarOutset;
