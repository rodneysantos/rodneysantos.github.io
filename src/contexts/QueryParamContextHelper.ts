/**
 * getURLSearchParams returns an array with the URI and URLSearchParams of the current location.
 * @returns {[string, URLSearchParams]} An array with the URI and URLSearchParams of the current location.
 */
export function getURLSearchParams(): [string, URLSearchParams] {
  const [uri, params] = window.location.href.split("?");
  return [uri, new URLSearchParams(params)];
}
