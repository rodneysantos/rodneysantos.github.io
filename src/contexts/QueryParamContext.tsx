import React, { createContext, useContext, useEffect, useState } from "react";
import { Keyword } from "../types";
import { getURLSearchParams } from "./QueryParamContextHelper";

interface QueryParams {
  photo: string;
  keywords: Keyword[];
}

interface QueryParamsContextProps {
  photo: string;
  keywords: Keyword[];
  setPhoto?: (photo: string) => void;
  toggleKeyword?: (keyword: Keyword) => void;
}

const QueryParamContext = createContext<QueryParamsContextProps>({
  photo: "",
  keywords: [],
});

const useQueryParams = () => useContext(QueryParamContext);

type WithQueryParams = <T>(
  Component: React.FC<T>,
  defaults?: QueryParams,
) => React.FC<T & {}>;

const withQueryParams: WithQueryParams =
  (Component, defaults = { photo: "", keywords: [] }) => (props) => {
    const [queryParams, setQueryParams] = useState<QueryParams>({
      photo: defaults.photo,
      keywords: defaults.keywords,
    });

    useEffect(() => {
      const [, searchParams] = getURLSearchParams();
      const keywordsParam = searchParams.get("keywords");

      if (keywordsParam === null) {
        // setting the default keyword so that when the page loads,
        // only B&W images are displayed.
        setURLSearchParams(defaults.keywords);
      } else {
        const keywords = keywordsParam.split(",") as Keyword[];
        setQueryParams({ ...queryParams, keywords });
        setURLSearchParams(keywords);
      }
    }, []);

    /**
     * setPhoto sets the photo query parameter in the URL and updates the corresponding state.
     * @param {string} photo - The photo to set in the URL.
     */
    function setPhoto(photo: string) {
      const [uri, searchParams] = getURLSearchParams();

      if (photo === "") {
        searchParams.delete("photo");
      } else {
        searchParams.set("photo", photo);
      }

      setQueryParams({ ...queryParams, photo });
      replaceHistoryState(uri, searchParams);
    }

    /**
     * toggleKeyword toggles the selection state of a keyword in the URL and updates the corresponding state.
     * @param {Keyword} keyword - The keyword to toggle the selection state for.
     */
    function toggleKeyword(keyword: Keyword) {
      const keywords = [...queryParams.keywords];

      if (keywords.includes(keyword)) {
        const index = keywords.indexOf(keyword);
        keywords.splice(index, 1);
      } else {
        keywords.push(keyword);
      }

      setURLSearchParams(keywords);
      setQueryParams({ ...queryParams, keywords });
    }

    return (
      <>
        <QueryParamContext.Provider
          value={{
            photo: queryParams.photo,
            keywords: queryParams.keywords,
            setPhoto,
            toggleKeyword,
          }}
        >
          <Component {...props} />
        </QueryParamContext.Provider>
      </>
    );

    /**
     * setURLSearchParams sets the keywords query parameter in the URL based on the provided keywords.
     * @param {Keyword[]} keywords - The keywords to set in the URL.
     */
    function setURLSearchParams(keywords: Keyword[]) {
      const [uri, searchParams] = getURLSearchParams();

      if (keywords.length > 0) {
        searchParams.set("keywords", keywords.join(","));
      } else {
        searchParams.delete("keywords");
      }

      replaceHistoryState(uri, searchParams);
    }

    /**
     * replaceHistoryState replaces the current history state with a new URL based on the provided URI and search parameters.
     * @param {string} uri - The URI to use in the new URL.
     * @param {URLSearchParams} params - The search parameters to use in the new URL.
     */
    function replaceHistoryState(uri: string, params: URLSearchParams) {
      const hasQueryParams = Array.from(params.values()).length > 0;
      const url = hasQueryParams ? `?${params.toString()}` : uri;
      window.history.replaceState(null, "", url);
    }
  };

export { QueryParams, useQueryParams, withQueryParams };
