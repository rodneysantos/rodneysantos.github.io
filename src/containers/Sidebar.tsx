import classNames from "classnames";
import React, { useEffect, useState } from "react";
import { ChevronIcon } from "../components/Icons";
import { useQueryParams } from "../contexts/QueryParamContext";
import * as icon from "../images/icons/bmc-full-logo.svg";
import { Keyword, Menu } from "../types";
import { useSidebarOutset } from "./SidebarOutset";

interface KeywordTag {
  isSelected: boolean;
  key: Keyword;
  name: string;
}

interface Navigation {
  key: Menu;
  name: string;
}

interface SidebarProps {
  keywordSelectedHandler: (keyword: Keyword) => void;
}

const Sidebar: React.FC<SidebarProps> = ({ keywordSelectedHandler }) => {
  const [keywords, setKeywords] = useState<KeywordTag[]>([
    { key: "architecture", name: "Architecture", isSelected: false },
    { key: "black-and-white", name: "Black and White", isSelected: false },
    { key: "color", name: "Color", isSelected: false },
    { key: "film-simulation", name: "Film Simulation", isSelected: false },
    { key: "low-key", name: "Low-Key", isSelected: false },
  ]);
  const queryParam = useQueryParams();
  const sidebarOutset = useSidebarOutset();
  const links: Navigation[] = [{ key: "about", name: "About" }];
  const cns = cn();

  useEffect(() => {
    const defaultKeywords = queryParam.keywords;

    if (defaultKeywords.length > 0) {
      setKeywords((outdatedKeywords) => {
        const updatedKeywords = outdatedKeywords.map((k) => {
          if (defaultKeywords.includes(k.key)) {
            return { ...k, isSelected: true };
          }

          return { ...k };
        });

        return updatedKeywords;
      });
    }
  }, []);

  function toggleSidebar() {
    sidebarOutset.setIsVisible!(!sidebarOutset.isVisible);
  }

  function toggleKeyword(keyword: KeywordTag) {
    setKeywords((outdatedKeywords) => {
      const updatedKeywords = outdatedKeywords.map((k) => ({ ...k }));
      const tag = updatedKeywords.find((k) => k.key === keyword.key)!;
      tag.isSelected = !tag.isSelected;

      return updatedKeywords;
    });

    keywordSelectedHandler(keyword.key as Keyword);
  }

  return (
    <>
      <div className="flex flex-col w-full h-screen bg-white">
        <div className={cns.brand}>
          <div className="leading-6">
            <span className="tracking-wider">RODNEY SANTOS</span>
            <br />
            <span className="font-dosis text-base tracking-[0.48rem]">
              PHOTOGRAPHY
            </span>
          </div>

          <ChevronIcon
            className={`h-6 relative ${{
              "rotate-180": sidebarOutset.isVisible,
            }} w-12`}
            clickHandler={toggleSidebar}
          />
        </div>

        <div className='flex flex-row h-screen'>
          <div className="flex flex-col flex-1 h-full overflow-hidden">
            <div className="mt-5 font-medium grow" data-testid="navigation">
              <ul className="leading-8 list-none" data-testid="menu">
                {links.map(({ key, name }) => (
                  <li
                    className="pl-8 cursor-pointer"
                    key={key}
                    data-testid={key}
                  >
                    {name}
                  </li>
                ))}

                <hr className="mx-8 my-3 border-slate-700" />

                {keywords.map((keyword) => (
                  <li
                    className={cns.keyword(keyword.isSelected)}
                    key={keyword.key}
                    data-testid={keyword.key}
                    onClick={() => toggleKeyword(keyword)}
                  >
                    {keyword.name}
                  </li>
                ))}
              </ul>
            </div>

            <a
              className={cns.support}
              href="https://www.buymeacoffee.com/xenon.photo"
              target="_blank"
              rel="noreferrer"
            >
              <img
                className="w-32"
                src={icon.default}
                alt="Buy me a coffee icon"
              />

              <span className="mt-1 text-xs text-center">
                Â© {new Date().getFullYear()}
              </span>
            </a>
          </div>
        </div>
      </div>
    </>
  );
};

function cn() {
  return {
    brand: classNames(
      "flex",
      "flex-row",
      "font-big-shoulder",
      "font-bold",
      "items-center",
      "pl-8",
      "pt-14",
      "text-3xl",
    ),
    keyword: (isSelected: boolean) =>
      classNames(
        { "bg-slate-900 px-2 text-slate-100": isSelected },
        "cursor-pointer",
        "mt-1",
        "mx-8",
        "select-none",
        "transition-all",
        "will-change-contents",
        "will-change-transform",
      ),
    support: classNames(
      "content-center",
      "flex",
      "flex-col",
      "flex-wrap",
      "pb-12",
      "pr-8",
      "w-full",
    ),
  };
}

export default Sidebar;
