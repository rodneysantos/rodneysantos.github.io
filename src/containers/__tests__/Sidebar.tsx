import { fireEvent, render } from "@testing-library/react";
import React from "react";
import rendeder from "react-test-renderer";
import { SidebarOutsetContext } from "../SidebarOutset";
import Sidebar from "../Sidebar";
import { withQueryParams } from "../../contexts/QueryParamContext";

describe("Sidebar component", () => {
  const menuTestID = "menu";
  const mockHandler = jest.fn();
  const url = "http://localhost/";
  let locationHrefSpy: jest.SpyInstance;

  beforeEach(() => {
    const url = "http://localhost";
    Object.defineProperty(window, "location", {
      value: new URL(url),
      configurable: true,
    });
    locationHrefSpy = jest.spyOn(window.location, "href", "get");
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly", () => {
    // arrange
    const tree = rendeder
      .create(<Sidebar keywordSelectedHandler={mockHandler} />)
      .toJSON();

    // assert
    expect(tree).toMatchSnapshot();
  });

  it("displays correct brand name", () => {
    // arrange
    const brandName = "RODNEY SANTOS";
    const { getByText } = render(
      <Sidebar keywordSelectedHandler={mockHandler} />,
    );

    // assert
    expect(getByText(brandName)).toBeDefined();
  });

  it("renders each item in the root menus array", () => {
    // arrange
    const { getAllByTestId } = render(
      <Sidebar keywordSelectedHandler={mockHandler} />,
    );

    // act
    const menus = getAllByTestId(menuTestID);

    // assert
    expect(menus.length).toEqual(1);
    expect(menus[0].children[0].textContent).toEqual("About");
    expect(menus[0].children[2].textContent).toEqual("Architecture");
  });

  it("toggles the sidebar when the brand icon is clicked", () => {
    // arrange
    const isVisible = false;
    const setIsVisible = jest.fn();
    const { getByTestId } = render(
      <SidebarOutsetContext.Provider value={{ isVisible, setIsVisible }}>
        <Sidebar keywordSelectedHandler={mockHandler} />
      </SidebarOutsetContext.Provider>,
    );

    // act
    const brandIcon = getByTestId("chevron-icon");
    fireEvent.click(brandIcon);

    // assert
    expect(setIsVisible).toHaveBeenCalledWith(true);
  });

  it("toggles a keyword when clicked", () => {
    // arrange
    const { getByTestId } = render(
      <Sidebar keywordSelectedHandler={mockHandler} />,
    );

    // act
    const colorKeyword = getByTestId("color");
    fireEvent.click(colorKeyword);

    // assert
    expect(mockHandler).toHaveBeenCalledWith("color");
    expect(colorKeyword.className).toContain("bg-slate-900");
  });

  it("highlights default keywords when provided", () => {
    // arrange
    const Component = withQueryParams(
      () => <Sidebar keywordSelectedHandler={mockHandler} />,
      { photo: "", keywords: ["color"] },
    );
    const { getByTestId } = render(<Component />);

    // act
    const colorKeyword = getByTestId("color");

    // assert
    expect(colorKeyword.className).toContain("bg-slate-900");
  });

  it("highlights the keywords that are not in the default list when provided", () => {
    // arrange
    locationHrefSpy.mockReturnValue(`${url}/?keywords=architecture%2Ccolor`);
    const Component = withQueryParams(
      () => <Sidebar keywordSelectedHandler={mockHandler} />,
      { photo: "", keywords: ["color"] },
    );
    const { getByTestId } = render(<Component />);

    // act
    const architectureKeyword = getByTestId("architecture");

    // assert
    expect(architectureKeyword.className).toContain("bg-slate-900");
  });
});
