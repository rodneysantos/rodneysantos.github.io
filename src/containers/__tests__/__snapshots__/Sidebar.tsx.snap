// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component renders correctly 1`] = `
<div
  className="bg-white flex flex-col h-screen w-full"
>
  <div
    className="flex flex-row font-big-shoulder font-bold items-center pl-8 pt-14 text-3xl"
  >
    <div
      className="leading-6"
    >
      <span
        className="tracking-wider"
      >
        RODNEY SANTOS
      </span>
      <br />
      <span
        className="font-dosis text-base tracking-[0.48rem]"
      >
        PHOTOGRAPHY
      </span>
    </div>
    <svg
      className="h-6 relative w-12"
      data-testid="chevron-icon"
      fill="currentColor"
      onClick={[Function]}
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clipRule="evenodd"
        d="M7.21 14.77a.75.75 0 01.02-1.06L11.168 10 7.23 6.29a.75.75 0 111.04-1.08l4.5 4.25a.75.75 0 010 1.08l-4.5 4.25a.75.75 0 01-1.06-.02z"
        fillRule="evenodd"
      />
    </svg>
  </div>
  <div
    className="flex flex-row h-screen"
  >
    <div
      className="flex flex-1 flex-col h-full overflow-hidden"
    >
      <div
        className="font-medium grow mt-5"
        data-testid="navigation"
      >
        <ul
          className="leading-8 list-none"
          data-testid="menu"
        >
          <li
            className="cursor-pointer pl-8"
            data-testid="about"
          >
            About
          </li>
          <hr
            className="border-slate-700 my-3 mx-8"
          />
          <li
            className="cursor-pointer mt-1 mx-8 select-none transition-all will-change-contents will-change-transform"
            data-testid="architecture"
            onClick={[Function]}
          >
            Architecture
          </li>
          <li
            className="cursor-pointer mt-1 mx-8 select-none transition-all will-change-contents will-change-transform"
            data-testid="black-and-white"
            onClick={[Function]}
          >
            Black and White
          </li>
          <li
            className="cursor-pointer mt-1 mx-8 select-none transition-all will-change-contents will-change-transform"
            data-testid="color"
            onClick={[Function]}
          >
            Color
          </li>
          <li
            className="cursor-pointer mt-1 mx-8 select-none transition-all will-change-contents will-change-transform"
            data-testid="film-simulation"
            onClick={[Function]}
          >
            Film Simulation
          </li>
          <li
            className="cursor-pointer mt-1 mx-8 select-none transition-all will-change-contents will-change-transform"
            data-testid="low-key"
            onClick={[Function]}
          >
            Low-Key
          </li>
        </ul>
      </div>
      <a
        className="content-center flex flex-col flex-wrap pb-12 pr-8 w-full"
        href="https://www.buymeacoffee.com/xenon.photo"
        rel="noreferrer"
        target="_blank"
      >
        <img
          alt="Buy me a coffee icon"
          className="w-32"
          src={{}}
        />
        <span
          className="mt-1 text-center text-xs"
        >
          Â© 
          2023
        </span>
      </a>
    </div>
  </div>
</div>
`;
